### ISSUE TITLE
Default output directory for `module-info.class` should be `META-INF/versions/9`

### ISSUE THREAD

**Issue by lukehutch** [moditect/moditect#67]
There are still a lot of old Java tools that choke when they hit a `module-info.class` file. It may take a long time for this problem to be solved, since old libraries pull in even older libraries in their transitive dependency graph.

`META-INF/versions/9` was the location recommended by Alan Bateman on the jigsaw-dev mailing list recently as the maximally backwards-compatible location for the module descriptor (the thread name is "Java Modules and Multi-Release Jar" -- Alan's message is not yet in the archive, but should be there in a few hours).

http://mail.openjdk.java.net/pipermail/jigsaw-dev/2018-September/thread.html#13932

Some tools (such as the Android toolchain) are being updated to ignore classes found in `META-INF/versions` for this reason:

https://issuetracker.google.com/u/0/issues/77587908

I tried using `<outputDirectory>` to specify where the module descriptor would be saved, but that affects the output jarfiles, not the location of `module-info.class`. Is there a way to save the module descriptor anywhere other than the root of the build target? If so, it would be a good idea to have the default be `target/META-INF/versions/9`, at least for the foreseeable future.

**Comment by gunnarmorling** [comment]
Can you try specifying the `jvmVersion` option as [described here](https://github.com/moditect/moditect#adding-a-module-descriptor-to-the-project-jar)? This should cause the descriptor to end up under META-INF/VERSIONS/... as you suggest.

**Comment by lukehutch** [comment]
@gunnarmorling Sorry, I totally missed that option! But I just tried it, and the descriptor was still put in the root of the build. I have:

```
<plugin>
	<groupId>org.moditect</groupId>
	<artifactId>moditect-maven-plugin</artifactId>
	<version>1.0.0.Beta1</version>
	<executions>
		<execution>
			<id>add-module-infos</id>
			<phase>package</phase>
			<goals>
				<goal>add-module-info</goal>
			</goals>
			<configuration>
			    <jvmVersion>9</jvmVersion>
				<overwriteExistingFiles>true</overwriteExistingFiles>
				<module>
					<moduleInfoFile>
						src/moditect/module-info.java
					</moduleInfoFile>
				</module>
			</configuration>
		</execution>
	</executions>
</plugin>
```

**Comment by lukehutch** [comment]
PS the docs are a bit hard to parse right now... this is not the first time I missed something important in the docs. I don't really know how I would format them differently though...

PPS Based on Alan's recommendation, maybe the default for `jvmVersion` should be 9? That actually always works, because (1) multi-release jars are supported on all JVMs that support JPMS; and (2) the version resolution order is the following:

http://mail.openjdk.java.net/pipermail/jigsaw-dev/2018-September/013935.html

This means that putting the module descriptor in `META-INF/versions/9` has the maximal chance of working with all future JDK versions, while staying out of the way of JDK 8 and earlier, by not polluting the "base layer" (the standard package root).

**Comment by lukehutch** [comment]
And actually the `jvmVersion` option is never shown in the debug logs for the ModiTect plugin:

```
[INFO] --- moditect-maven-plugin:1.0.0.Beta1:add-module-info (add-module-infos) @ classgraph ---
[DEBUG] Dependency collection stats: {ConflictMarker.analyzeTime=31311, ConflictMarker.markTime=23721, ConflictMarker.nodeCount=6, ConflictIdSorter.graphTime=11140, ConflictIdSorter.topsortTime=19021, ConflictIdSorter.conflictIdCount=6, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=180725, ConflictResolver.conflictItemCount=6, DefaultDependencyCollector.collectTime=6688860, DefaultDependencyCollector.transformTime=379780}
[DEBUG] org.moditect:moditect-maven-plugin:jar:1.0.0.Beta1:
[DEBUG]    org.eclipse.aether:aether-util:jar:1.0.2.v20150114:compile
[DEBUG]       org.eclipse.aether:aether-api:jar:1.0.2.v20150114:compile
[DEBUG]    org.moditect:moditect:jar:1.0.0.Beta1:compile
[DEBUG]       com.github.javaparser:javaparser-core:jar:3.5.0:compile
[DEBUG]       com.beust:jcommander:jar:1.60:compile
[DEBUG]    org.codehaus.plexus:plexus-utils:jar:1.1:runtime
[DEBUG] Created new class realm plugin>org.moditect:moditect-maven-plugin:1.0.0.Beta1
[DEBUG] Importing foreign packages into class realm plugin>org.moditect:moditect-maven-plugin:1.0.0.Beta1
[DEBUG]   Imported:  < project>io.github.classgraph:classgraph:4.2.6-SNAPSHOT
[DEBUG] Populating class realm plugin>org.moditect:moditect-maven-plugin:1.0.0.Beta1
[DEBUG]   Included: org.moditect:moditect-maven-plugin:jar:1.0.0.Beta1
[DEBUG]   Included: org.eclipse.aether:aether-util:jar:1.0.2.v20150114
[DEBUG]   Included: org.moditect:moditect:jar:1.0.0.Beta1
[DEBUG]   Included: com.github.javaparser:javaparser-core:jar:3.5.0
[DEBUG]   Included: com.beust:jcommander:jar:1.60
[DEBUG]   Included: org.codehaus.plexus:plexus-utils:jar:1.1
[DEBUG] Configuring mojo org.moditect:moditect-maven-plugin:1.0.0.Beta1:add-module-info from plugin realm ClassRealm[plugin>org.moditect:moditect-maven-plugin:1.0.0.Beta1, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@799f7e29]
[DEBUG] Configuring mojo 'org.moditect:moditect-maven-plugin:1.0.0.Beta1:add-module-info' with basic configurator -->
[DEBUG]   (f) artifactId = classgraph
[DEBUG]   (f) buildDirectory = /home/luke/Work/classgraph/target
[DEBUG]   (s) moduleInfoFile = /home/luke/Work/classgraph/src/moditect/module-info.java
[DEBUG]   (f) module = MainModuleConfiguration [ moduleInfo=null, moduleInfoFile=/home/luke/Work/classgraph/src/moditect/module-info.java, moduleInfoSource=null, mainClass=null]
[DEBUG]   (f) outputDirectory = /home/luke/Work/classgraph/target/modules
[DEBUG]   (f) overwriteExistingFiles = true
[DEBUG]   (f) project = MavenProject: io.github.classgraph:classgraph:4.2.6-SNAPSHOT @ /home/luke/Work/classgraph/pom.xml
[DEBUG]   (f) remoteRepos = [jboss (https://repository.jboss.org/, default, releases), central (http://jcenter.bintray.com, default, releases)]
[DEBUG]   (f) repoSession = org.eclipse.aether.DefaultRepositorySystemSession@5d9d8e46
[DEBUG]   (f) version = 4.2.6-SNAPSHOT
[DEBUG]   (f) workingDirectory = /home/luke/Work/classgraph/target/moditect
[DEBUG] -- end configuration --```

**Comment by gunnarmorling** [comment]
Ah, just realizing this feature is part of Beta2, which hasn't been
released yet. I can try and cut a release soon, until then you'd have to
build ModiTect from source in order to use this.

**Comment by lukehutch** [comment]
Ah, OK. No worries, I'll wait. Please consider the suggestion to default this option to 9 though. I know it seems weird on the surface to put just that one file into a version directory, but that provides maximum compatibility.

**Comment by gunnarmorling** [comment]
Hum, hum. It seems a bit weird on first thought indeed, but perhaps it's really best compat-wise. The question then is how to use that option to put the file into root if really wanted. Perhaps `<jvmVersion>NONE</jvmVersion>`?

**Comment by gunnarmorling** [comment]
> PS the docs are a bit hard to parse right now... this is not the first time I missed something important in the docs. I don't really know how I would format them differently though...

I hear you. Open for any suggestions for re-organizing them. Or better yet, a PR :)

**Comment by lukehutch** [comment]
Yes, I think `NONE` is fine.

The issues I have hit with a few users of my library (and I have hit it myself) is that older classpath scanners will throw an NPE when they run into the module descriptor. It's not nessicarily true that older scanners will ignore classfiles in META-INF, but the thinking is that that's the best chance you have at putting classfiles somewhere that an older scanner might intentionally not scan. (And some scanners ignore classfiles whose packages don't match their path, which hopefully happens before the NPE or similar is thrown).

**Comment by lukehutch** [comment]
Plus it's the location of the classfile we're talking about, not the Java source, so it doesn't matter if the target location is not where most humans would expect it to be, as long as it is the most robust location.

**Comment by gunnarmorling** [comment]
Resolved via #68.

**Comment by lukehutch** [comment]
In case anyone is referring back to this issue at some point, I found direct support for this in [JEP 238](http://openjdk.java.net/jeps/238):

> A multi-release modular need not have a module descriptor at the located root. In this respect a module descriptor would be treated no differently to any other class or resource file. This can ensure that, for example, only Java 8 versioned classes are present in the root area while Java 9 versioned classes (including the module descriptor) are present in the 9 versioned area.

**Comment by lukehutch** [comment]
@gunnarmorling Any idea when you plan to release Beta2? I have a user who has been hitting up against problems using ClassGraph because the Android build tools can't handle `module-info.class` in the root directory of a jarfile...

**Comment by lukehutch** [comment]
@gunnarmorling Any chance you can please push out a new release? I got yet another report of problems from third party runtime environments dying when they find a module-info.class in the root of the jar. (This time it's Tomcat BCEL, called by Apache Catalina.) This is probably the third or fourth time I have run into this. I could manually move the module-info.class into a versioned section of the jar, but I'd rather automate it with the new Moditect features. Thanks!

**Comment by gunnarmorling** [comment]
Hey @lukehutch, thanks for the reminder! Apologies for the delay (and just ignoring your earlier question about the release, I somehow missed that). I'll try and push out a release over the weekend.

**Comment by lukehutch** [comment]
Thank you!

**Comment by gunnarmorling** [comment]
Agh, so I ran into some issues with the set-up for the release today :( I'll keep working on this in the course of the next days.
